batch/batch/worker/worker.py

```diff
@@ -366,6 +367,68 @@ async def _free(self, netns: NetworkNamespace):
             self.public_networks.put_nowait(netns)
 
 
+class FuseMount:
+    def __init__(self, path):
+        self.path = path
+        self.bind_mounts = set()
+
+
+# Mounts that can be shared across jobs by the same user
+# Only sharing within jobs of the same user ensures that
+# the user is authorized to access the bucket
+class ReadOnlyCloudfuseManager:
+    def __init__(self):
+        self.cloudfuse_dir = '/cloudfuse/readonly_cache'
+        self.fuse_mounts: Dict[Tuple[str, str], FuseMount] = {}
+        self.user_bucket_locks: Dict[Tuple[str, str], asyncio.Lock] = defaultdict(asyncio.Lock)
+
+    async def mount(
+        self, bucket: str, destination: str, *, user: str, credentials_path: str, tmp_path: str, config: dict
+    ):
+        assert config['read_only']
+        async with self.user_bucket_locks[(user, bucket)]:
+            if (user, bucket) not in self.fuse_mounts:
+                local_path = self._new_path()
+                await self._fuse_mount(local_path, credentials_path=credentials_path, tmp_path=tmp_path, config=config)
+                self.fuse_mounts[(user, bucket)] = FuseMount(local_path)
+            mount = self.fuse_mounts[(user, bucket)]
+            await self._bind_mount(mount.path, destination)
+            mount.bind_mounts.add(destination)
+
+    async def unmount(self, destination, *, user: str, bucket: str):
+        async with self.user_bucket_locks[(user, bucket)]:
+            mount = self.fuse_mounts[(user, bucket)]
+            await self._bind_unmount(destination)
+            mount.bind_mounts.remove(destination)
+            if len(mount.bind_mounts) == 0:
+                await self._fuse_unmount(mount.path)
+                del self.fuse_mounts[(user, bucket)]
+
+    async def _fuse_mount(self, destination: str, *, credentials_path: str, tmp_path: str, config: dict):
+        assert CLOUD_WORKER_API
+        await CLOUD_WORKER_API.mount_cloudfuse(
+            credentials_path,
+            destination,
+            tmp_path,
+            config,
+        )
+
+    async def _fuse_unmount(self, path: str):
+        assert CLOUD_WORKER_API
+        await CLOUD_WORKER_API.unmount_cloudfuse(path)
+
+    async def _bind_mount(self, src, dst):
+        await check_exec_output('mount', '--bind', src, dst)
```

Comment:
Just to double check I understand why there is now an additional `mount` and `umount`, this is because we have the additional level of nesting where as with the regular DockerJob, the mounts were handled by `crun` when it created the container?

Reply:
Yes exactly! On `JVMContainer` creation, we create the `/cloudfuse` mountpoint and then these `mounts/umount`s are underneath that and get propagated into the running container.